File: matlab/src/lang/layered/Task.m
classdef Task < LayeredNetworkElement
    % A software server in a LayeredNetwork.
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.
    
    properties
        parent;
        multiplicity;       %int
        replication;       %int
        scheduling;         %string
        priority = 0;       %int, priority level (0 = default/no priority)
        fanInSource = '';   %string, source task for fan-in
        fanInValue = 0;     %int, fan-in value (load distribution count)
        thinkTime;
        thinkTimeMean;      %double
        thinkTimeSCV;       %double
        setupTime;
        setupTimeMean;      %double
        setupTimeSCV;       %double
        delayOffTime;
        delayOffTimeMean;   %double
        delayOffTimeSCV;    %double
        entries = [];
        activities = [];
        precedences = [];
        replyEntry;
    end
    
    
    methods
        %public methods, including constructor
        
        %constructor
        function self = Task(model, name, multiplicity, scheduling, thinkTime)
            % self = TASK(MODEL, NAME, MULTIPLICITY, SCHEDULING, THINKTIME)
            name = char(name);
            if nargin<2%~exist('name','var')
                line_error(mfilename,'Constructor requires to specify at least a name.');
            end
            self@LayeredNetworkElement(name);
            
            if nargin<3%~exist('multiplicity','var')
                multiplicity = 1;
            end
            if nargin<4%~exist('scheduling','var')
                scheduling = SchedStrategy.INF;
            end
            if nargin<5%~exist('thinkTime','var')
                thinkTime = GlobalConstants.FineTol;
            end
            self.replication = 1;
            self.multiplicity = multiplicity;
            switch scheduling
                case SchedStrategy.INF
                if isfinite(multiplicity)
                    line_warning(mfilename,'Finite multiplicity is not allowed with INF scheduling. Setting it to INF.\n');
                    self.multiplicity = Inf;
                end
            end
            self.scheduling = SchedStrategy.toText(scheduling);
            self.setThinkTime(thinkTime);
            self.setSetupTime(Immediate());
            self.setDelayOffTime(Immediate());
            self.parent = [];

            if isa(model,'LayeredNetwork')
                model.tasks{end+1} = self;
                switch scheduling
                    case 'ref'
                        model.reftasks{end+1} = self;
                end            
                self.model = model;
            elseif isa(model,'JLayeredNetwork')
                % JLayeredNetwork support would go here if it exists
                self.obj = jline.lang.layered.Task(model.obj, name, multiplicity, scheduling, thinkTime);
                model.addTask(self);
            end
        end
        
        function self = setReplication(self, replication)
            self.replication = replication;
        end
                
        function self = on(self, parent)
            % self = ON(self, PARENT)
            if ~isa(parent,'Host') && ~isa(parent,'Processor')
                line_error(mfilename,'Invalid .on() argument: expected a host processor.')
            end            
            if isempty(self.parent)
                self.parent = parent;
                parent.addTask(self);
            else
                line_error(mfilename,'Parent processor already defined.')
            end
        end
        
        function self = setAsReferenceTask(self)
            % self = SETASREFERENCETASK(self)
            
            self.scheduling = SchedStrategy.REF;
        end
        
        function self = setThinkTime(self, thinkTime)
            % self = SETTHINKTIME(self, THINKTIME)

            if isnumeric(thinkTime)
                if thinkTime <= GlobalConstants.FineTol
                    self.thinkTime = Immediate.getInstance();
                    self.thinkTimeMean = GlobalConstants.FineTol;
                    self.thinkTimeSCV = GlobalConstants.FineTol;
                else
                    self.thinkTime = Exp(1/thinkTime);
                    self.thinkTimeMean = thinkTime;
                    self.thinkTimeSCV = 1.0;
                end
            elseif isa(thinkTime,'Distribution')
                self.thinkTime = thinkTime;
                self.thinkTimeMean = thinkTime.getMean();
                self.thinkTimeSCV = thinkTime.getSCV();
            end
        end

        function self = setSetupTime(self, setupTime)
            % self = SETSETUPTIME(self, SETUPTIME)
            % Set the setup time (cold start time) for the task.

            if isnumeric(setupTime)
                if setupTime <= GlobalConstants.FineTol
                    self.setupTime = Immediate.getInstance();
                    self.setupTimeMean = GlobalConstants.FineTol;
                    self.setupTimeSCV = GlobalConstants.FineTol;
                else
                    self.setupTime = Exp(1/setupTime);
                    self.setupTimeMean = setupTime;
                    self.setupTimeSCV = 1.0;
                end
            elseif isa(setupTime,'Distribution')
                self.setupTime = setupTime;
                self.setupTimeMean = setupTime.getMean();
                self.setupTimeSCV = setupTime.getSCV();
            end
        end

        function self = setDelayOffTime(self, delayOffTime)
            % self = SETDELAYOFFTIME(self, DELAYOFFTIME)
            % Set the delay-off time (teardown time) for the task.

            if isnumeric(delayOffTime)
                if delayOffTime <= GlobalConstants.FineTol
                    self.delayOffTime = Immediate.getInstance();
                    self.delayOffTimeMean = GlobalConstants.FineTol;
                    self.delayOffTimeSCV = GlobalConstants.FineTol;
                else
                    self.delayOffTime = Exp(1/delayOffTime);
                    self.delayOffTimeMean = delayOffTime;
                    self.delayOffTimeSCV = 1.0;
                end
            elseif isa(delayOffTime,'Distribution')
                self.delayOffTime = delayOffTime;
                self.delayOffTimeMean = delayOffTime.getMean();
                self.delayOffTimeSCV = delayOffTime.getSCV();
            end
        end

        function result = hasSetupDelayoff(self)
            % result = HASSETUPDELAYOFF(self)
            % Check if this task has setup/delayoff configured (i.e., non-trivial values).

            hasSetup = ~isempty(self.setupTime) && ~isa(self.setupTime, 'Immediate') ...
                       && self.setupTimeMean > GlobalConstants.FineTol;
            hasDelayoff = ~isempty(self.delayOffTime) && ~isa(self.delayOffTime, 'Immediate') ...
                          && self.delayOffTimeMean > GlobalConstants.FineTol;
            result = hasSetup || hasDelayoff;
        end

        %addEntry
        function self = addEntry(self, newEntry)
            % self = ADDENTRY(self, NEWENTRY)
            
            self.entries = [self.entries; newEntry];
        end
        
        %addActivity
        function self = addActivity(self, newAct)
            % self = ADDACTIVITY(self, NEWACT)
            
            newAct.setParent(self.name);
            self.activities = [self.activities; newAct];
        end
        
        %setActivity
        function self = setActivity(self, newAct, index)
            % self = SETACTIVITY(self, NEWACT, INDEX)
            
            self.activities(index,1) = newAct;
        end
        
        %removeActivity
        function self = removeActivity(self, index)
            % self = REMOVEACTIVITY(self, INDEX)
            
            idxToKeep = [1:index-1,index+1:length(self.activities)];
            self.activities = self.activities(idxToKeep);
            self.actNames = self.actNames(idxToKeep);
        end
        
        %addPrecedence
        function self = addPrecedence(self, newPrec)
            % self = ADDPRECEDENCE(self, NEWPREC)
            
            if iscell(newPrec)
                for m=1:length(newPrec)
                    self.precedences = [self.precedences; newPrec{m}];
                end
            else
                self.precedences = [self.precedences; newPrec];
            end
        end
        
        %setReplyEntry
        function self = setReplyEntry(self, newReplyEntry)
            % self = SETREPLYENTRY(self, NEWREPLYENTRY)
            
            self.replyEntry = newReplyEntry;
        end
        
        function meanHostDemand = getMeanHostDemand(self, entryName)
            % MEANHOSTDEMAND = GETMEANHOSTDEMAND(self, ENTRYNAME)

            % determines the demand posed by the entry entryName
            % the demand is located in the activity of the corresponding entry

            meanHostDemand = -1;
            for j = 1:length(self.entries)
                if strcmp(self.entries(j).name, entryName)
                    meanHostDemand = self.entries(j).activities(1).hostDemandMean;
                    break;
                end
            end
        end

        % Getter methods for API consistency with Java/Python
        function val = getMultiplicity(obj)
            % GETMULTIPLICITY Get the multiplicity (number of task instances)
            val = obj.multiplicity;
        end

        function val = getReplication(obj)
            % GETREPLICATION Get the replication factor
            val = obj.replication;
        end

        function val = getScheduling(obj)
            % GETSCHEDULING Get the scheduling strategy
            val = obj.scheduling;
        end

        function val = getThinkTimeMean(obj)
            % GETTHINKTIMEMEAN Get the mean think time
            val = obj.thinkTimeMean;
        end

        function val = getThinkTimeSCV(obj)
            % GETTHINKTIMESCV Get the SCV of think time
            val = obj.thinkTimeSCV;
        end

        function val = getParent(obj)
            % GETPARENT Get the parent host/processor
            val = obj.parent;
        end

        function val = getPrecedences(obj)
            % GETPRECEDENCES Get the list of activity precedences
            val = obj.precedences;
        end

        function val = getSetupTimeMean(obj)
            % GETSETUPTIMEMEAN Get the mean setup time
            val = obj.setupTimeMean;
        end

        function val = getDelayOffTimeMean(obj)
            % GETDELAYOFFTIMEMEAN Get the mean delay-off time
            val = obj.delayOffTimeMean;
        end

        % setPriority
        function self = setPriority(self, priority)
            % self = SETPRIORITY(self, PRIORITY)
            self.priority = priority;
        end

        % setFanIn
        function self = setFanIn(self, source, value)
            % self = SETFANIN(self, SOURCE, VALUE)
            self.fanInSource = source;
            self.fanInValue = value;
        end

    end

end

File: jar/src/main/kotlin/jline/lang/layered/Task.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.layered;

import jline.GlobalConstants;
import jline.lang.constant.SchedStrategy;
import jline.lang.processes.Distribution;
import jline.lang.processes.Exp;
import jline.lang.processes.Immediate;

import java.util.ArrayList;
import java.util.List;

import static jline.io.InputOutputKt.line_error;
import static jline.io.InputOutputKt.mfilename;

/**
 * A Task represents a software component or process in a layered queueing network that can host
 * services (Entry objects) and runs on a Processor (Host).
 * 
 * <p>Tasks are fundamental building blocks in layered queueing networks that encapsulate:
 * <ul>
 * <li>Service entries that define the interfaces accessible to other tasks</li>
 * <li>Activities that represent internal processing steps</li>
 * <li>Activity precedences that define the execution order and relationships</li>
 * <li>Resource demands and scheduling policies</li>
 * </ul>
 * 
 * <p>Key characteristics:
 * <ul>
 * <li><b>Multiplicity:</b> Number of concurrent task instances that can execute</li>
 * <li><b>Scheduling:</b> How the task schedules its internal processing (FCFS, PS, etc.)</li>
 * <li><b>Think time:</b> Time between completing one request and starting the next</li>
 * <li><b>Reference tasks:</b> Special tasks that generate workload (external customers)</li>
 * </ul>
 * 
 * <p>Tasks can be reference tasks (SchedStrategy.REF) that represent external workload
 * generators, or regular tasks that provide services to other tasks in the system.
 * 
 * @see Entry
 * @see Activity
 * @see ActivityPrecedence
 * @see Processor
 * @see LayeredNetwork
 */
public class Task extends LayeredNetworkElement {
    protected Processor parent;
    protected int multiplicity;
    protected int replication;
    protected SchedStrategy scheduling;//Enum
    protected int priority = 0;  // Priority level (0 = default/no priority)
    protected String fanInSource = "";  // Source task for fan-in
    protected int fanInValue = 0;  // Fan-in value (load distribution count)
    protected Distribution thinkTime;
    protected double thinkTimeMean;
    protected double thinkTimeSCV;
    protected Distribution setupTime;
    protected double setupTimeMean;
    protected double setupTimeSCV;
    protected Distribution delayOffTime;
    protected double delayOffTimeMean;
    protected double delayOffTimeSCV;
    protected List<Entry> entries;
    protected List<Activity> activities;
    protected List<ActivityPrecedence> precedences;
    private List<Entry> replyEntry;

    public Task(LayeredNetwork model, String name, int multiplicity, SchedStrategy scheduling, Distribution thinkTime) {
        super(name);
        this.parent = null;
        this.setReplication(1);
        this.model = model;
        this.multiplicity = multiplicity;
        this.scheduling = scheduling;
        this.setThinkTime(thinkTime);
        this.setSetupTime(Immediate.getInstance());
        this.setDelayOffTime(Immediate.getInstance());
        this.entries = new ArrayList<>();
        this.activities = new ArrayList<>();
        this.precedences = new ArrayList<>();
        this.replyEntry = new ArrayList<>();
        // link within model
        model.tasks.put(model.tasks.size(), this);
        model.nodes.put(model.tasks.size(), this);
        if (scheduling == SchedStrategy.REF) {
            model.reftasks.put(model.reftasks.size(), this);
        }
    }

    public Task(LayeredNetwork model, String name, int multiplicity, SchedStrategy scheduling) {
        this(model, name, multiplicity, scheduling, Immediate.getInstance());
    }

    public Task(LayeredNetwork model, String name, int multiplicity) {
        this(model, name, multiplicity, SchedStrategy.INF, Immediate.getInstance());
    }

    public Task(LayeredNetwork model, String name) {
        this(model, name, 1, SchedStrategy.INF, Immediate.getInstance());
    }

    public Task addActivity(Activity newActivity) {
        newActivity.setParent(this);
        this.activities.add(newActivity);
        return this;
    }

    public Task addEntry(Entry newEntry) {
        this.entries.add(newEntry);
        return this;
    }

    public Task addPrecedence(ActivityPrecedence newPrec) {
        this.precedences.add(newPrec);
        return this;
    }

    public Task addPrecedence(List<ActivityPrecedence> newPrec) {
        this.precedences.addAll(newPrec);
        return this;
    }

    public Task addPrecedence(ActivityPrecedence[] newPrec) {
        for (ActivityPrecedence ap : newPrec) {
            this.precedences.add(ap);
        }
        return this;
    }

    public double getMeanHostDemand(String entryName) {
        double meanHostDemand = -1;
        for (int j = 0; j < this.entries.size(); j++) {
            if (this.entries.get(j).getName().equals(entryName)) {
                break;
            }
        }
        return meanHostDemand;
    }

    public Task on(Processor parent) {
        if (this.parent != null) {
            line_error(mfilename(new Object() {
            }), "Parent processor already defined.");
        }
        this.parent = parent;
        this.parent.addTask(this);
        return this;
    }

    public Task removeActivity(int index) {
        this.activities.remove(index);
        return this;
    }

    public Task setActivity(Activity newActivity, int index) {
        this.activities.set(index, newActivity);
        return this;
    }

    public Task setAsReferenceTask() {
        this.scheduling = SchedStrategy.REF;
        return this;
    }

    public Task setReplication(int replication) {
        this.replication = replication;
        return this;
    }

    public Task setReplyEntry(List<Entry> replyEntry) {
        this.replyEntry = replyEntry;
        return this;
    }

    public Task setThinkTime(Distribution thinkTime) {
        this.thinkTime = thinkTime;
        this.thinkTimeMean = thinkTime.getMean();
        this.thinkTimeSCV = thinkTime.getSCV();
        return this;
    }

    public Task setThinkTime(double thinkTime) {
        if (thinkTime <= GlobalConstants.Zero) {
            this.thinkTime = Immediate.getInstance();
            this.thinkTimeMean = GlobalConstants.Zero;
            this.thinkTimeSCV = GlobalConstants.Zero;
        } else {
            this.thinkTime = new Exp(1 / thinkTime);
            this.thinkTimeMean = thinkTime;
            this.thinkTimeSCV = 1.0;
        }
        return this;
    }

    /**
     * Set the setup time (cold start time) for the task.
     *
     * @param setupTime The setup time distribution
     * @return This Task instance for method chaining
     */
    public Task setSetupTime(Distribution setupTime) {
        this.setupTime = setupTime;
        this.setupTimeMean = setupTime.getMean();
        this.setupTimeSCV = setupTime.getSCV();
        return this;
    }

    /**
     * Set the setup time using a mean value (creates exponential distribution).
     *
     * @param setupTime The mean setup time
     * @return This Task instance for method chaining
     */
    public Task setSetupTime(double setupTime) {
        if (setupTime <= GlobalConstants.FineTol) {
            this.setupTime = Immediate.getInstance();
            this.setupTimeMean = GlobalConstants.FineTol;
            this.setupTimeSCV = GlobalConstants.FineTol;
        } else {
            this.setupTime = new Exp(1.0 / setupTime);
            this.setupTimeMean = setupTime;
            this.setupTimeSCV = 1.0;
        }
        return this;
    }

    /**
     * Set the delay-off time (teardown time) for the task.
     *
     * @param delayOffTime The delay-off time distribution
     * @return This Task instance for method chaining
     */
    public Task setDelayOffTime(Distribution delayOffTime) {
        this.delayOffTime = delayOffTime;
        this.delayOffTimeMean = delayOffTime.getMean();
        this.delayOffTimeSCV = delayOffTime.getSCV();
        return this;
    }

    /**
     * Set the delay-off time using a mean value (creates exponential distribution).
     *
     * @param delayOffTime The mean delay-off time
     * @return This Task instance for method chaining
     */
    public Task setDelayOffTime(double delayOffTime) {
        if (delayOffTime <= GlobalConstants.FineTol) {
            this.delayOffTime = Immediate.getInstance();
            this.delayOffTimeMean = GlobalConstants.FineTol;
            this.delayOffTimeSCV = GlobalConstants.FineTol;
        } else {
            this.delayOffTime = new Exp(1.0 / delayOffTime);
            this.delayOffTimeMean = delayOffTime;
            this.delayOffTimeSCV = 1.0;
        }
        return this;
    }

    // Getters for setup time
    public Distribution getSetupTime() {
        return setupTime;
    }

    public double getSetupTimeMean() {
        return setupTimeMean;
    }

    public double getSetupTimeSCV() {
        return setupTimeSCV;
    }

    // Getters for delay-off time
    public Distribution getDelayOffTime() {
        return delayOffTime;
    }

    public double getDelayOffTimeMean() {
        return delayOffTimeMean;
    }

    public double getDelayOffTimeSCV() {
        return delayOffTimeSCV;
    }

    public int getMultiplicity() {
        return multiplicity;
    }

    public int getReplication() {
        return replication;
    }

    public SchedStrategy getScheduling() {
        return scheduling;
    }

    public double getThinkTimeMean() {
        return thinkTimeMean;
    }

    public double getThinkTimeSCV() {
        return thinkTimeSCV;
    }

    public Host getParent() {
        return parent;
    }

    public void setMultiplicity(int multiplicity) {
        this.multiplicity = multiplicity;
    }

    public void setScheduling(SchedStrategy scheduling) {
        this.scheduling = scheduling;
    }

    public List<ActivityPrecedence> getPrecedences() {
        return precedences;
    }

    /**
     * Check if this task has setup/delayoff configured (i.e., non-trivial values).
     *
     * @return true if setup time or delay-off time is configured with non-immediate values
     */
    public boolean hasSetupDelayoff() {
        boolean hasSetup = setupTime != null && !(setupTime instanceof Immediate)
                           && setupTimeMean > GlobalConstants.FineTol;
        boolean hasDelayoff = delayOffTime != null && !(delayOffTime instanceof Immediate)
                              && delayOffTimeMean > GlobalConstants.FineTol;
        return hasSetup || hasDelayoff;
    }

    /**
     * Returns the list of entries for this task.
     *
     * @return the list of entries
     */
    public List<Entry> getEntries() {
        return entries;
    }

    /**
     * Returns the list of activities for this task.
     *
     * @return the list of activities
     */
    public List<Activity> getActivities() {
        return activities;
    }

    /**
     * Returns the processor (host) that this task runs on.
     *
     * @return the parent processor
     */
    public Processor getProcessor() {
        return parent;
    }

    public Task setPriority(int priority) {
        this.priority = priority;
        return this;
    }

    public int getPriority() {
        return this.priority;
    }

    public Task setFanIn(String source, int value) {
        this.fanInSource = source;
        this.fanInValue = value;
        return this;
    }

    public String getFanInSource() {
        return this.fanInSource;
    }

    public int getFanInValue() {
        return this.fanInValue;
    }

}
